File: symmetric-tree/symmetric-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        
        def check(nodeLeft, nodeRight):
            if not nodeLeft and not nodeRight:
                return True
            
            if not nodeLeft or not nodeRight:
                return False
            
            return (nodeLeft.val == nodeRight.val) and check(nodeLeft.left, nodeRight.right) and check(nodeLeft.right, nodeRight.left)
        
        
        return check(root, root)

File: symmetric-tree/symmetric-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isSymmetric(TreeNode root) {
        return check(root, root);
    }
    
    public boolean check(TreeNode nodeLeft, TreeNode nodeRight) {
        if (nodeLeft == null && nodeRight == null) {
            return true;
        }
        
        if (nodeLeft == null || nodeRight == null) {
            return false;
        }
        
        return (nodeLeft.val == nodeRight.val) && check(nodeLeft.left, nodeRight.right) && check(nodeLeft.right, nodeRight.left);
    }
}